tests/test_numpy_wrapper.py

```diff
@@ -370,11 +370,77 @@ def cost(x):
 
     def test_non_differentiable_gradient(self):
         """Test gradient computation with requires_grad=False raises an error"""
+
         def cost(x):
             return np.sum(np.sin(x))
 
         grad_fn = qml.grad(cost, argnum=[0])
-        arr1 = np.array([0., 1., 2.], requires_grad=False)
+        arr1 = np.array([0.0, 1.0, 2.0], requires_grad=False)
 
         with pytest.raises(np.NonDifferentiableError, match="non-differentiable"):
             grad_fn(arr1)
+
+
+class TestScalarHashing:
+    """Test for the hashing capability of scalar arrays."""
+
+    def test_create_set_scalar_arrays(self):
+        """Test that a collection of scalar arrays can be properly hashed
+        when creating a set"""
+        data = [np.array(1), np.array(2), np.array(1), np.array(3)]
+        res = set(data)
+        expected = {np.array(1), np.array(2), np.array(3)}
+        assert res == expected
+
+    def test_requires_grad_hashing(self):
+        """Test that the gradient information is correctly taken into account when hashing"""
+        data = [np.array(1), np.array(2), np.array(1, requires_grad=False), np.array(3)]
+        res = set(data)
+        expected = {np.array(1, requires_grad=False), np.array(1), np.array(2), np.array(3)}
+        assert res == expected
+
+    def test_create_set_from_array_iteration(self):
+        """Test that a one dimensional array correctly produces a set via iteration"""
+        data = np.array([1, 2, 1, 3], requires_grad=True)
+        res = set(data)
+        expected = {
+            np.array(1, requires_grad=True),
+            np.array(2, requires_grad=True),
+            np.array(3, requires_grad=True),
+        }
+        assert res == expected
+
+        data = np.array([1, 2, 1, 3], requires_grad=False)
+        res = set(data)
+        expected = {
+            np.array(1, requires_grad=False),
+            np.array(2, requires_grad=False),
+            np.array(3, requires_grad=False),
+        }
+        assert res == expected
+
+    def test_nonzero_dim_arrays_non_hashable(self):
+        """Test that a non-scalar array continues to remain non-hashable"""
+        with pytest.raises(TypeError, match=r"unhashable type: 'numpy\.tensor'"):
+            set(np.array([[1, 2], [3, 4]]))
+
+
+class TestNumpyConversion:
+    """Tests for the tensor.numpy() method"""
+
+    def test_convert_scalar_array(self):
+        """Test that a scalar array converts to a python literal"""
+        data = np.array(1.543)
+        res = data.numpy()
+        assert res == data.item()
+        assert isinstance(res, float)
+
+    def test_convert_array(self):
+        """Test that a numpy array successfully converts"""
+        data = np.array([1, 2, 3])
+        res = data.numpy()
+
+        assert np.shares_memory(res, data)
+        assert np.all(res == data)
+        assert isinstance(res, np.ndarray)
+        assert not isinstance(res, np.tensor)

```

Comment:
Do we want to add a test that checks that `numpy()` returns a reference (just to keep this behaviour fixed)?

Reply:
That's currently being done by the line
```python
assert np.shares_memory(res, data)
```